### PR TITLE
Use only nats for occs_nums rather than ints

### PR DESCRIPTION
**Pull request by jfehrle**
In https://github.com/coq/coq/pull/13343#discussion_r524341931, @Zimmi48 noticed that `change a with b in * at -1 -2` is accepted, or `... at -1 2` but `... at 1 -2` is not.  That's because of the way that the `occs_nums` productions are written:

```
occs_nums: [
| LIST1 nat_or_var
| "-" nat_or_var LIST0 int_or_var
]
```

This change replaces the `int_or_var`.  The line taking the absolute value of the ints is from 2008 by one @herbelin.  Perhaps after 12 years we can remove this?  OTOH there hasn't been a deprecation message.

Please advise whether I should add a deprecation message first.  Also whether this should have a change log entry.

### CODE REVIEW

**Comment by herbelin** [comment]
> Please advise whether I should add a deprecation message first. Also whether this should have a change log entry.

I see no occurrences of `at -n -p` in the Coq contributions, so I guess that there is no risk at removing it. And with a change log we would be on the very safe side.

File: doc/tools/docgram/orderedGrammar [hunk 1]
```diff
@@ -648,7 +648,7 @@ ref_or_pattern_occ: [
 
 occs_nums: [
 | LIST1 [ natural | ident ]
-| "-" [ natural | ident ] LIST0 int_or_var
+| "-" LIST1 [ natural | ident ]
```
Comment (herbelin) [inline doc/tools/docgram/orderedGrammar:651]: Not that the `nat_or_var` expresses an idea different than `[ natural | ident ]`. The `nat_or_var` indicates that it is a natural or an Ltac variable denoting a natural. With `[ natural | ident ]`, isn't there a risk that one thinks at a natural or an hypothesis?

I know that the ambiguity already exists in other situations (e.g. we don't write `hyp_or_var` to mean an hypothesis or an Ltac variable denoting an hypothesis, since it is an `ident` in both cases), so maybe this is hopeless anyway.

Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar:651]: I'm open to eventually creating more nonterminals that just to show the semantics better, such as `var` (in both the doc and the mlgs), but that's not my priority now.  The text in the doc should make clear that the ident is for a variable.

**Comment by Zimmi48** [comment]
I've changed the milestone: 8.13+beta1 can still be used until the release if @gares grants the inclusion. It's a bit on the edge in this case, but it would also be fine to delay this to 8.14. In any case, it wouldn't be appropriate as a change between the beta and the final release (which 8.13.0 would imply).

File: doc/tools/docgram/orderedGrammar (hunk 1 above)
Comment (Zimmi48) [reply doc/tools/docgram/orderedGrammar:651]: Interesting, this corresponds to my suggestion in https://github.com/coq/coq/pull/13343#discussion_r523794967.
@jfehrle: Your answer to this suggestion said that the `nat_or_var` nonterminal exists but was little used so spliced in the documentation. IMHO it would then be worth to:
1. use it in this case (no semantics change)
2. not splice it in the documentation (for symmetry with the `int_or_var` nonterminal).

Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar:651]: No, it corresponds to https://github.com/coq/coq/pull/13343#discussion_r524341931, which I mentioned in the PR description.  Splicing or not splicing `nat_or_var` is an independent issue better handled where it was raised in #13343.

**Comment by jfehrle** [comment]
I added a changelog entry.  Unless you or @Zimmi48 have further comments, I think this is ready to merge.  Perhaps @herbelin is willing to be the assignee?

@gares Can we include this in 8.13?  Not a big deal if you'd rather not.

File: doc/tools/docgram/fullGrammar [hunk 2]
```diff
@@ -604,7 +605,7 @@ command: [
 | "Typeclasses" "Opaque" LIST1 reference
 | "Typeclasses" "eauto" ":=" debug eauto_search_strategy OPT natural
 | "Proof" "with" Pltac.tactic OPT [ "using" G_vernac.section_subset_expr ]
-| "Proof" "using" G_vernac.section_subset_expr OPT [ "with" Pltac.tactic ]
+| "Proof" "using" G_vernac.section_subset_expr "with" Pltac.tactic
```
Comment (herbelin) [inline doc/tools/docgram/fullGrammar:608]: I don't care so much but the `Proof using` changes are unrelated to this PR (these are obviously late changes to the `Proof using` PR #13339).

File: doc/tools/docgram/orderedGrammar [hunk 3]
```diff
@@ -1960,6 +1961,7 @@ or_and_intropattern_loc: [
 
 as_or_and_ipat: [
 | "as" or_and_intropattern_loc
+| "as" equality_intropattern
```
Comment (herbelin) [inline doc/tools/docgram/orderedGrammar:1964]: This extra line is only for generating an error message at parsing time. I would say that it should not be documented (even if there is a discussion at #13205).

File: doc/sphinx/proofs/writing-proofs/rewriting.rst [hunk 4]
```diff
@@ -320,7 +320,7 @@ Performing computations
    ref_or_pattern_occ ::= @reference {? at @occs_nums }
    | @one_term {? at @occs_nums }
    occs_nums ::= {+ {| @natural | @ident } }
-   | - {| @natural | @ident } {* @int_or_var }
+   | - {+ {| @natural | @ident } }
```
Comment (herbelin) [inline doc/sphinx/proofs/writing-proofs/rewriting.rst:323]: Didn't you intended to write `nat_or_var` here and in `orderedGrammar`, as it is in `fullGrammar`?

**Review by herbelin — APPROVED** [review]
Good to me up to remarks. Assigning.

File: doc/tools/docgram/fullGrammar (hunk 2 above)
Comment (Zimmi48) [reply doc/tools/docgram/fullGrammar:608]: Yeah, but until we implement a CI check that `fullGrammar` and `orderGrammar` are updated, this is how we update these files.

File: doc/changelog/04-tactics/13403-occs_nums_nat.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+- **Removed:**
+  "at @occ_nums" clauses in tactics such as autounfold
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/13403-occs_nums_nat.rst:2]: I would not use `autounfold` as an example since this tactic is little used. What about using `unfold` instead as an example?

```suggestion
  :n:`at @occ_nums` clauses in tactics such as :tacn:`autounfold`
```

File: doc/changelog/04-tactics/13403-occs_nums_nat.rst [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  no longer allow negative values.  A "-" before the
+  list (for set complement) is still supported.  Ex: "at 1 -2"
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/13403-occs_nums_nat.rst:4]: This example is bad because it was never supported. `at -1 -2` was but it meant the same thing as `at -1 2`.

File: doc/changelog/04-tactics/13403-occs_nums_nat.rst [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+  is no longer supported but "at -1 2" is.  Does not
+  affect rewrite * or setoid_rewrite.
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/13403-occs_nums_nat.rst:6]: As far as I understand `rewrite` and `setoid_rewrite` have *exactly* the same issue, so I suggest restricting the syntax accepted by `rewrite` and `setoid_rewrite` in the same way. But even if you do not implement that, this changelog should not mention these tactics anyway because we should certainly not advertise this problematic syntax that we should remove eventually.

File: doc/sphinx/proofs/writing-proofs/rewriting.rst (hunk 4 above)
Comment (Zimmi48) [reply doc/sphinx/proofs/writing-proofs/rewriting.rst:323]: For now, this nonterminal is "spliced". I am in favor of not splicing it anymore, but Jim said that we should address this question separately as part as a documentation PR like #13343.

File: doc/tools/docgram/fullGrammar (hunk 2 above)
Comment (jfehrle) [reply doc/tools/docgram/fullGrammar:608]: `make doc_gram_rsts` automatically generates `fullGrammar` from all the mlgs.  In this case, we're picking up an mlg change from another PR for which the submitter didn't run `doc_grammar`.  `fullGrammar` shouldn't be edited by hand.  It is for reference, doesn't affect the binaries.

File: doc/tools/docgram/orderedGrammar (hunk 3 above)
Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar:1964]: Sure.  `as_or_and_ipat` is in the tactics chapter, which hasn't been updated yet.  We'll handle it then.

File: doc/sphinx/proofs/writing-proofs/rewriting.rst (hunk 4 above)
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/rewriting.rst:323]: I already changed this in #13343.  This change was automatically generated by doc_grammar based on `fullGrammar` and the editing instructions in `common.edit_mlg` , BTW.

File: doc/changelog/04-tactics/13403-occs_nums_nat.rst (hunk 7 above)
Comment (jfehrle) [reply doc/changelog/04-tactics/13403-occs_nums_nat.rst:6]: It's not exactly the same.  Those two use different nonterminals to specify the occurrences and slightly different code.  #13391 recommends changing them so they use the same nonterminal  and have the same functionality as `unfold` et al.  Tweaking `occurrences` and function `occurrences_of` in the mlgs is a band aid when something more is needed.  (Note that I rename mlg `occurrences` to `rewrite_occs` in the doc and similarly `clause_dft_concl` in the mlg to `occurrences` in the doc.)

I removed the sentence.

Comment (jfehrle) [reply doc/changelog/04-tactics/13403-occs_nums_nat.rst:6]: FWIW, I found the following tactics have `int_or_var` where it appears not to be useful (some give error messages, others don't):
constructor, econstructor, do, timeout, hresolve_core, hget_var, auto, info_auto, eauto, info_eauto, dfs eauto, probably bfs eauto, probably typeclasses eauto, psatz_Z, psatz_R, psatz_Q, psatz

debug_auto and debug_eauto give slightly different results for negative numbers, but negatives don't make sense.

I'll create a separate PR for those.

**Comment by jfehrle** [comment]
Updated.

**Comment by gares** [comment]
> @gares Can we include this in 8.13? Not a big deal if you'd rather not.

All documentation related PR have backport granted (unless they conflict with each others).

**Comment by Zimmi48** [comment]
It's not pure doc (I wouldn't backport it after the beta for instance) but it's indeed doc-oriented since all it removes is a useless accidental syntax. The same is true for #13417 BTW.

**Comment by gares** [comment]
I see it touches a grammar rule, but AFAICT it is aestetic (parses the same language). Am I wrong?

**Comment by Zimmi48** [comment]
No, it does remove some previously valid syntax: `at -1 -2 3` and `at -1 2 -3` were previously accepted (but meant the same thing as `at -1 2 3` and were not documented). This is described in the changelog entry.

**Comment by gares** [comment]
Ah ok, thanks for clarifying that. Then I'm OK with it as long as it comes in before the beta.

**Comment by herbelin** [comment]
Is this ready? If yes, I'll merge when requested (or, @gares, you can merge it directly, that'll be as simple).

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Yes, it's ready. @herbelin: please merge ASAP

**Comment by herbelin** [comment]
Metacoq still running and elpi failing are unrelated I guess.

@coqbot: merge now

**Comment by jfehrle** [comment]
Thanks for the review.
